projects/plugins/jetpack/modules/carousel/jetpack-carousel.php

```diff
@@ -284,27 +291,74 @@ class_exists( 'Jetpack_AMP_Support' )
 	}
 
 	/**
-	 * Check if the content of a post uses gallery blocks. To be used by 'the_content' filter.
+	 * Enrich the gallery block content using the render_block_{$this->name} filter.
+	 * This function is triggered after block render to make sure we track galleries within
+	 * reusable blocks.
 	 *
-	 * @since 6.8.0
+	 * @see https://developer.wordpress.org/reference/hooks/render_block_this-name/
 	 *
-	 * @param string $content Post content.
+	 * @param string $block_content The rendered HTML for the carousel or gallery block.
+	 * @param array  $block         The parsed block details for the block.
+	 * @return string The fully-processed HTML for the carousel or gallery block.
 	 *
-	 * @return string $content Post content.
+	 * @since $$next-release$$
 	 */
-	public function check_content_for_blocks( $content ) {
-		if (
-			class_exists( 'Jetpack_AMP_Support' )
-			&& Jetpack_AMP_Support::is_amp_request()
-		) {
-			return $content;
+	public function filter_gallery_block_render( $block_content, $block ) {
+		global $post;
+
+		if ( empty( $block['blockName'] ) || ! in_array( $block['blockName'], array( 'core/gallery', 'jetpack/tiled-gallery' ), true ) ) {
+			return $block_content;
 		}
 
-		if ( has_block( 'gallery', $content ) || has_block( 'jetpack/tiled-gallery', $content ) ) {
-			$this->enqueue_assets();
-			$content = $this->add_data_to_container( $content );
+		$this->enqueue_assets();
+
+		if ( ! isset( $post ) ) {
+			return $block_content;
 		}
-		return $content;
+
+		$blog_id = (int) get_current_blog_id();
+
+		$extra_data = array(
+			'data-carousel-extra' => array(
+				'blog_id'   => $blog_id,
+				'permalink' => get_permalink( $post->ID ),
+			),
+		);
+
+		/**
+		 * Filter the data added to the Gallery container.
+		 *
+		 * @module carousel
+		 *
+		 * @since 1.6.0
+		 *
+		 * @param array $extra_data Array of data about the site and the post.
+		 */
+		$extra_data = apply_filters( 'jp_carousel_add_data_to_container', $extra_data );
+		$extra_data = (array) $extra_data;
+
+		if ( empty( $extra_data ) ) {
+			return $block_content;
+		}
+
+		$extra_attributes = implode(
+			' ',
+			array_map(
+				function ( $data_key, $data_values ) {
+					return esc_attr( $data_key ) . "='" . wp_json_encode( $data_values ) . "'";
+				},
+				array_keys( $extra_data ),
+				array_values( $extra_data )
+			)
+		);
+
+		// Add extra attributes to first HTML element (which may have leading whitespace)
+		return preg_replace(
+			'/^(\s*<(div|ul|figure))/',
```

Comment:
Can you confirm in what scenario there would be an `ul` root element? I can see `div` is used for shortcodes and `figure` seems to be used for two gallery blocks. Maybe there is a legacy gallery using `ul`?

Reply:
TBH, this is mostly something that in the existing code -- I wanted to keep it rather than assume whether it could or couldn't appear.

Reply:
ok, fair enough!